Bottom: cf3945762ae7656f74d823af7746002b238077e1
Top:    b5ba01a6f86e9daad8f40e783f9e7df6d63a19fb
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-12-23 17:19:44 +0000

Fix broken undo handling for ResetFound*() and ResetConnections()

Many callers have this idiom:

  SaveUndoSerialNumber ();
  ResetFoundPinsViasAndPads (AndDraw);
  RestoreUndoSerialNumber ();
  ResetFoundLinesAndPolygons (AndDraw);

The intent is to squash the two operations into one undo operation. The
ResetFound* functions DO increment the Undo serial number but ONLY if
that particular function makes a change.

Assume for the example that the first ResetFound* call makes some change
and increments the serial number. This is then reset by the caller. If
the second ResetFound* call doesn't find anything, there is no overall
increment to the serial number.

Move the undo serial number handling out of the ResetFound* calls, and
let the callers increment the serial number if desired. To facilitate
this, return a boolean value from these functions to indicate whether
any changes were made which could be undone.

Add an additional argument to the Reset* functions separating conditional
execution of drawing updates and undo related processing.

-----

Expose ResetConnections() API and use to simplify existing code

For convenience, expose previously static find.c function
ResetConnections() in order to avoid repeating Undo handling in each
caller to the two ResetFound* functions which are often called together.

-----

Remove redundant IncrementSerialNumber() from report.c

The ResetFound*() functions already increment the undo serial number
when called with true as an argument, and any changes are made.

---

diff --git a/src/action.c b/src/action.c
index 7a9c3f0..fde1e7c 100644
--- a/src/action.c
+++ b/src/action.c
@@ -2362,18 +2362,18 @@ ActionConnection (int argc, char **argv, int x, int y)
 	  }
 
 	case F_ResetLinesAndPolygons:
-	  ResetFoundLinesAndPolygons (true);
+	  if (ResetFoundLinesAndPolygons (true, true))
+	    IncrementUndoSerialNumber ();
 	  break;
 
 	case F_ResetPinsViasAndPads:
-	  ResetFoundPinsViasAndPads (true);
+	  if (ResetFoundPinsViasAndPads (true, true))
+	    IncrementUndoSerialNumber ();
 	  break;
 
 	case F_Reset:
-	  SaveUndoSerialNumber ();
-	  ResetFoundPinsViasAndPads (true);
-	  RestoreUndoSerialNumber ();
-	  ResetFoundLinesAndPolygons (true);
+	  if (ResetConnections (true, true))
+	    IncrementUndoSerialNumber ();
 	  break;
 	}
       RestoreCrosshair (true);
@@ -2786,10 +2786,8 @@ ActionDisplay (int argc, char **argv, int childX, int childY)
 	  TOGGLE_FLAG (AUTODRCFLAG, PCB);
 	  if (TEST_FLAG (AUTODRCFLAG, PCB) && Settings.Mode == LINE_MODE)
 	    {
-	      SaveUndoSerialNumber ();
-	      ResetFoundPinsViasAndPads (true);
-	      RestoreUndoSerialNumber ();
-	      ResetFoundLinesAndPolygons (true);
+	      if (ResetConnections (true, true))
+		IncrementUndoSerialNumber ();
 	      if (Crosshair.AttachedLine.State != STATE_FIRST)
 		LookupConnection (Crosshair.AttachedLine.Point1.X,
 				  Crosshair.AttachedLine.Point1.Y, true, 1,
diff --git a/src/autoroute.c b/src/autoroute.c
index 9ccab97..2c86270 100644
--- a/src/autoroute.c
+++ b/src/autoroute.c
@@ -985,8 +985,7 @@ CreateRouteData ()
    * this saves on searching the trees to find the nets
    */
   /* use the DRCFLAG to mark objects as they are entered */
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
+  ResetConnections (false, false);
   Nets = CollectSubnets (false);
   {
     routebox_t *last_net = NULL;
diff --git a/src/find.c b/src/find.c
index 3176357..8229868 100644
--- a/src/find.c
+++ b/src/find.c
@@ -345,7 +345,6 @@ static void PrintPinConnections (FILE *, bool);
 static bool PrintAndSelectUnusedPinsAndPadsOfElement (ElementTypePtr,
                                                          FILE *);
 static void DrawNewConnections (void);
-static void ResetConnections (bool);
 static void DumpList (void);
 static void LocateError (LocationType *, LocationType *);
 static void BuildObjectList (int *, long int **, int **);
@@ -3221,7 +3220,7 @@ LookupElementConnections (ElementTypePtr Element, FILE * FP)
   /* reset all currently marked connections */
   User = true;
   TheFlag = FOUNDFLAG;
-  ResetConnections (true);
+  ResetConnections (true, true);
   InitConnectionLookup ();
   PrintElementConnections (Element, FP, true);
   SetChangedFlag (true);
@@ -3242,7 +3241,7 @@ LookupConnectionsToAllElements (FILE * FP)
   /* reset all currently marked connections */
   User = false;
   TheFlag = FOUNDFLAG;
-  ResetConnections (false);
+  ResetConnections (false, false);
   InitConnectionLookup ();
 
   ELEMENT_LOOP (PCB->Data);
@@ -3252,12 +3251,12 @@ LookupConnectionsToAllElements (FILE * FP)
       break;
     SEPARATE (FP);
     if (Settings.ResetAfterElement && n != 1)
-      ResetConnections (false);
+      ResetConnections (false, false);
   }
   END_LOOP;
   if (Settings.RingBellWhenFinished)
     gui->beep ();
-  ResetConnections (false);
+  ResetConnections (false, false);
   FreeConnectionLookupMemory ();
   ClearAndRedrawOutput ();
 }
@@ -3419,7 +3418,7 @@ LookupUnusedPins (FILE * FP)
   /* reset all currently marked connections */
   User = true;
   SaveUndoSerialNumber ();
-  ResetConnections (true);
+  ResetConnections (true, true);
   RestoreUndoSerialNumber ();
   InitConnectionLookup ();
 
@@ -3444,20 +3443,19 @@ LookupUnusedPins (FILE * FP)
 /* ---------------------------------------------------------------------------
  * resets all used flags of pins and vias
  */
-void
-ResetFoundPinsViasAndPads (bool AndDraw)
+bool
+ResetFoundPinsViasAndPads (bool save_undo, bool redraw)
 {
   bool change = false;
 
-
   VIA_LOOP (PCB->Data);
   {
     if (TEST_FLAG (TheFlag, via))
       {
-        if (AndDraw)
+        if (save_undo)
           AddObjectToFlagUndoList (VIA_TYPE, via, via, via);
         CLEAR_FLAG (TheFlag, via);
-        if (AndDraw)
+        if (redraw)
           DrawVia (via, 0);
         change = true;
       }
@@ -3469,10 +3467,10 @@ ResetFoundPinsViasAndPads (bool AndDraw)
     {
       if (TEST_FLAG (TheFlag, pin))
         {
-          if (AndDraw)
+          if (save_undo)
             AddObjectToFlagUndoList (PIN_TYPE, element, pin, pin);
           CLEAR_FLAG (TheFlag, pin);
-          if (AndDraw)
+          if (redraw)
             DrawPin (pin, 0);
           change = true;
         }
@@ -3482,10 +3480,10 @@ ResetFoundPinsViasAndPads (bool AndDraw)
     {
       if (TEST_FLAG (TheFlag, pad))
         {
-          if (AndDraw)
+          if (save_undo)
             AddObjectToFlagUndoList (PAD_TYPE, element, pad, pad);
           CLEAR_FLAG (TheFlag, pad);
-          if (AndDraw)
+          if (redraw)
             DrawPad (pad, 0);
           change = true;
         }
@@ -3496,31 +3494,28 @@ ResetFoundPinsViasAndPads (bool AndDraw)
   if (change)
     {
       SetChangedFlag (true);
-      if (AndDraw)
-        {
-          IncrementUndoSerialNumber ();
-          Draw ();
-        }
+      if (redraw)
+        Draw ();
     }
+  return change;
 }
 
 /* ---------------------------------------------------------------------------
  * resets all used flags of LOs
  */
-void
-ResetFoundLinesAndPolygons (bool AndDraw)
+bool
+ResetFoundLinesAndPolygons (bool save_undo, bool redraw)
 {
   bool change = false;
 
-
   RAT_LOOP (PCB->Data);
   {
     if (TEST_FLAG (TheFlag, line))
       {
-        if (AndDraw)
+        if (save_undo)
           AddObjectToFlagUndoList (RATLINE_TYPE, line, line, line);
         CLEAR_FLAG (TheFlag, line);
-        if (AndDraw)
+        if (redraw)
           DrawRat (line, 0);
         change = true;
       }
@@ -3530,10 +3525,10 @@ ResetFoundLinesAndPolygons (bool AndDraw)
   {
     if (TEST_FLAG (TheFlag, line))
       {
-        if (AndDraw)
+        if (save_undo)
           AddObjectToFlagUndoList (LINE_TYPE, layer, line, line);
         CLEAR_FLAG (TheFlag, line);
-        if (AndDraw)
+        if (redraw)
           DrawLine (layer, line, 0);
         change = true;
       }
@@ -3543,10 +3538,10 @@ ResetFoundLinesAndPolygons (bool AndDraw)
   {
     if (TEST_FLAG (TheFlag, arc))
       {
-        if (AndDraw)
+        if (save_undo)
           AddObjectToFlagUndoList (ARC_TYPE, layer, arc, arc);
         CLEAR_FLAG (TheFlag, arc);
-        if (AndDraw)
+        if (redraw)
           DrawArc (layer, arc, 0);
         change = true;
       }
@@ -3556,10 +3551,10 @@ ResetFoundLinesAndPolygons (bool AndDraw)
   {
     if (TEST_FLAG (TheFlag, polygon))
       {
-        if (AndDraw)
+        if (save_undo)
           AddObjectToFlagUndoList (POLYGON_TYPE, layer, polygon, polygon);
         CLEAR_FLAG (TheFlag, polygon);
-        if (AndDraw)
+        if (redraw)
           DrawPolygon (layer, polygon, 0);
         change = true;
       }
@@ -3568,26 +3563,27 @@ ResetFoundLinesAndPolygons (bool AndDraw)
   if (change)
     {
       SetChangedFlag (true);
-      if (AndDraw)
-        {
-          IncrementUndoSerialNumber ();
-          Draw ();
-        }
+      if (redraw)
+        Draw ();
     }
+  return change;
 }
 
 /* ---------------------------------------------------------------------------
  * resets all found connections
  */
-static void
-ResetConnections (bool AndDraw)
+bool
+ResetConnections (bool save_undo, bool redraw)
 {
-  if (AndDraw)
-    SaveUndoSerialNumber ();
-  ResetFoundPinsViasAndPads (AndDraw);
-  if (AndDraw)
-    RestoreUndoSerialNumber ();
-  ResetFoundLinesAndPolygons (AndDraw);
+  bool change = false;
+
+  change = ResetFoundPinsViasAndPads  (save_undo, redraw) || change;
+  change = ResetFoundLinesAndPolygons (save_undo, redraw) || change;
+
+  if (change && save_undo)
+    IncrementUndoSerialNumber ();
+
+  return change;
 }
 
 /*----------------------------------------------------------------------------
@@ -3657,7 +3653,7 @@ DRCFind (int What, void *ptr1, void *ptr2, void *ptr3)
           DumpList ();
           /* make the flag changes undoable */
           TheFlag = FOUNDFLAG | SELECTEDFLAG;
-          ResetConnections (false);
+          ResetConnections (false, false);
           User = true;
           drc = false;
           Bloat = -PCB->Shrink;
@@ -3706,7 +3702,7 @@ DRCFind (int What, void *ptr1, void *ptr2, void *ptr3)
     }
   /* now check the bloated condition */
   drc = false;
-  ResetConnections (false);
+  ResetConnections (false, false);
   TheFlag = FOUNDFLAG;
   ListStart (What, ptr1, ptr2, ptr3);
   Bloat = PCB->Bloat;
@@ -3717,7 +3713,7 @@ DRCFind (int What, void *ptr1, void *ptr2, void *ptr3)
       DumpList ();
       /* make the flag changes undoable */
       TheFlag = FOUNDFLAG | SELECTEDFLAG;
-      ResetConnections (false);
+      ResetConnections (false, false);
       User = true;
       drc = false;
       Bloat = 0;
@@ -3775,7 +3771,7 @@ DRCFind (int What, void *ptr1, void *ptr2, void *ptr3)
   drc = false;
   DumpList ();
   TheFlag = FOUNDFLAG | SELECTEDFLAG;
-  ResetConnections (false);
+  ResetConnections (false, false);
   return (false);
 }
 
@@ -3934,7 +3930,7 @@ DRCAll (void)
 
   TheFlag = FOUNDFLAG | DRCFLAG | SELECTEDFLAG;
 
-  ResetConnections (true);
+  ResetConnections (true, true);
 
   User = false;
 
@@ -3984,7 +3980,7 @@ DRCAll (void)
   END_LOOP;
 
   TheFlag = (IsBad) ? DRCFLAG : (FOUNDFLAG | DRCFLAG | SELECTEDFLAG);
-  ResetConnections (false);
+  ResetConnections (false, false);
   TheFlag = SELECTEDFLAG;
   /* check minimum widths and polygon clearances */
   if (!IsBad)
diff --git a/src/find.h b/src/find.h
index 5b04980..5b92248 100644
--- a/src/find.h
+++ b/src/find.h
@@ -54,8 +54,9 @@ void LookupElementConnections (ElementTypePtr, FILE *);
 void LookupConnectionsToAllElements (FILE *);
 void LookupConnection (LocationType, LocationType, bool, BDimension, int);
 void LookupUnusedPins (FILE *);
-void ResetFoundLinesAndPolygons (bool);
-void ResetFoundPinsViasAndPads (bool);
+bool ResetFoundLinesAndPolygons (bool save_undo, bool redraw);
+bool ResetFoundPinsViasAndPads (bool save_undo, bool redraw);
+bool ResetConnections (bool save_undo, bool redraw);
 void InitConnectionLookup (void);
 void InitComponentLookup (void);
 void InitLayoutLookup (void);
diff --git a/src/hid/gtk/gui-netlist-window.c b/src/hid/gtk/gui-netlist-window.c
index cee63da..6e129cc 100644
--- a/src/hid/gtk/gui-netlist-window.c
+++ b/src/hid/gtk/gui-netlist-window.c
@@ -516,17 +516,16 @@ netlist_select_cb (GtkWidget * widget, gpointer data)
     node_selected_net = NULL;
 
   InitConnectionLookup ();
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
-  SaveUndoSerialNumber ();
+  ResetConnections (false, false);
 
+  SaveUndoSerialNumber ();
   for (i = selected_net->EntryN, entry = selected_net->Entry; i; i--, entry++)
     if (SeekPad (entry, &conn, false))
       RatFindHook (conn.type, conn.ptr1, conn.ptr2, conn.ptr2, true, true);
   RestoreUndoSerialNumber ();
+
   SelectConnection (select_flag);
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
+  ResetConnections (false, false);
   FreeConnectionLookupMemory ();
   IncrementUndoSerialNumber ();
   Draw ();
diff --git a/src/hid/lesstif/netlist.c b/src/hid/lesstif/netlist.c
index 3438b33..b2952d8 100644
--- a/src/hid/lesstif/netlist.c
+++ b/src/hid/lesstif/netlist.c
@@ -136,8 +136,7 @@ nbcb_select_common (LibraryMenuTypePtr net, int pos, int select_flag)
   int i;
 
   InitConnectionLookup ();
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
+  ResetConnections (false);
   SaveUndoSerialNumber ();
 
   for (i = net->EntryN, entry = net->Entry; i; i--, entry++)
@@ -145,8 +144,7 @@ nbcb_select_common (LibraryMenuTypePtr net, int pos, int select_flag)
       RatFindHook (conn.type, conn.ptr1, conn.ptr2, conn.ptr2, true, true);
   RestoreUndoSerialNumber ();
   SelectConnection (select_flag);
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
+  ResetConncetions (false);
   FreeConnectionLookupMemory ();
   IncrementUndoSerialNumber ();
   Draw ();
diff --git a/src/rats.c b/src/rats.c
index 5fe86cd..3f2a8fa 100644
--- a/src/rats.c
+++ b/src/rats.c
@@ -463,8 +463,7 @@ GatherSubnets (NetListTypePtr Netl, bool NoWarn, bool AndRats)
   for (m = 0; Netl->NetN > 0 && m < Netl->NetN; m++)
     {
       a = &Netl->Net[m];
-      ResetFoundPinsViasAndPads (false);
-      ResetFoundLinesAndPolygons (false);
+      ResetConnections (false, false);
       RatFindHook (a->Connection[0].type, a->Connection[0].ptr1,
 		   a->Connection[0].ptr2, a->Connection[0].ptr2, false,
 		   AndRats);
@@ -545,8 +544,7 @@ GatherSubnets (NetListTypePtr Netl, bool NoWarn, bool AndRats)
       if (!NoWarn)
 	Warned |= CheckShorts (a->Connection[0].menu);
     }
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
+  ResetConnections (false, false);
   return (Warned);
 }
 
diff --git a/src/report.c b/src/report.c
index a9c429d..7d9063a 100644
--- a/src/report.c
+++ b/src/report.c
@@ -688,11 +688,8 @@ ReportAllNetLengths (int argc, char **argv, int x, int y)
       continue;
 
     got_one:
-      SaveUndoSerialNumber ();
-      ResetFoundPinsViasAndPads (true);
-      RestoreUndoSerialNumber ();
-      ResetFoundLinesAndPolygons (true);
-      RestoreUndoSerialNumber ();
+      ResetConnections (true, true);
+      /* FIXME: Need a serial number increment?? */
       length = XYtoNetLength (x, y, &found);
 
       gui->log("Net %s length %.*f %s\n", netname, prec, length*scale, units_name);
@@ -707,11 +704,8 @@ ReportNetLength (int argc, char **argv, int x, int y)
   char *netname = 0;
   int found = 0;
 
-  SaveUndoSerialNumber ();
-  ResetFoundPinsViasAndPads (true);
-  RestoreUndoSerialNumber ();
-  ResetFoundLinesAndPolygons (true);
-  RestoreUndoSerialNumber ();
+  ResetConnections (true, true);
+  /* FIXME: Need a serial number increment?? */
   gui->get_coords ("Click on a connection", &x, &y);
 
   length = XYtoNetLength (x, y, &found);
diff --git a/src/set.c b/src/set.c
index 7a41f73..837b786 100644
--- a/src/set.c
+++ b/src/set.c
@@ -292,11 +292,8 @@ SetMode (int Mode)
       Crosshair.AttachedLine.State = STATE_FIRST;
       if (Mode == LINE_MODE && TEST_FLAG (AUTODRCFLAG, PCB))
 	{
-	  SaveUndoSerialNumber ();
-	  ResetFoundPinsViasAndPads (true);
-	  RestoreUndoSerialNumber ();
-	  ResetFoundLinesAndPolygons (true);
-	  IncrementUndoSerialNumber ();
+	  if (ResetConnections (true, true))
+	    IncrementUndoSerialNumber ();
 	}
     }
 
diff --git a/src/toporouter.c b/src/toporouter.c
index 88c1325..7b44ec4 100644
--- a/src/toporouter.c
+++ b/src/toporouter.c
@@ -2799,8 +2799,7 @@ void
 import_clusters(toporouter_t *r)
 {
   NetListListType nets;
-  ResetFoundPinsViasAndPads (false);
-  ResetFoundLinesAndPolygons (false);
+  ResetConnections (false, false);
   nets = CollectSubnets(false);
   NETLIST_LOOP(&nets);
   {
